server/schedulers/hot_region.go

```diff
@@ -606,184 +672,264 @@ func (bs *balanceSolver) getDstCandidateIDs() map[uint64]struct{} {
 	}
 
-	if err != nil {
-		log.Debug("fail to create operator", zap.Error(err), zap.Stringer("opType", bs.opTy), zap.Stringer("rwType", bs.rwTy))
-		schedulerCounter.WithLabelValues(bs.sche.GetName(), "create-operator-fail").Inc()
-		return nil
-	}
+	if bs.cur.srcPeerStat != old.srcPeerStat {
+		// compare region
 
-	op.SetPriorityLevel(core.HighPriority)
-	op.Counters = append(op.Counters,
-		schedulerCounter.WithLabelValues(bs.sche.GetName(), "new-operator"),
-		schedulerCounter.WithLabelValues(bs.sche.GetName(), bs.opTy.String()))
+		if bs.rwTy == write && bs.opTy == transferLeader {
+			switch {
+			case bs.cur.srcPeerStat.GetKeyRate() > old.srcPeerStat.GetKeyRate():
+				return true
+			case bs.cur.srcPeerStat.GetKeyRate() < old.srcPeerStat.GetKeyRate():
+				return false
+			}
+		} else {
+			byteRkCmp := rankCmp(bs.cur.srcPeerStat.GetByteRate(), old.srcPeerStat.GetByteRate(), stepRank(0, 100))
+			keyRkCmp := rankCmp(bs.cur.srcPeerStat.GetKeyRate(), old.srcPeerStat.GetKeyRate(), stepRank(0, 10))
 
-	infl := Influence{ByteRate: bs.srcPeerStat.GetByteRate()}
-	if bs.opTy == transferLeader && bs.rwTy == write {
-		infl.ByteRate = 0
+			switch bs.cur.progressiveRank {
+			case -2: // 0.95 < byteDecRatio <= 0.99 && keyDecRatio <= 0.95
+				if keyRkCmp != 0 {
+					return keyRkCmp > 0
+				}
+				if byteRkCmp != 0 {
+					// prefer smaller byte rate, to reduce oscillation
+					return byteRkCmp < 0
+				}
+			case -3: // byteDecRatio <= 0.95 && keyDecRatio <= 0.95
+				if keyRkCmp != 0 {
+					return keyRkCmp > 0
+				}
+				fallthrough
+			case -1: // byteDecRatio <= 0.95
+				if byteRkCmp != 0 {
+					// prefer region with larger byte rate, to converge faster
+					return byteRkCmp > 0
+				}
+			}
+		}
 	}
-	bs.sche.addPendingInfluence(op, bs.srcStoreID, bs.dstStoreID, infl, bs.rwTy, bs.opTy)
 
-	return []*operator.Operator{op}
+	return false
 }
 
-// Sort stores according to their load prediction.
-func sortStores(loadDetail map[uint64]*storeLoadDetail, better func(lp1, lp2 *storeLoadPred) bool) []uint64 {
-	ids := make([]uint64, 0, len(loadDetail))
-	for id := range loadDetail {
-		ids = append(ids, id)
-	}
-	sort.Slice(ids, func(i, j int) bool {
-		id1, id2 := ids[i], ids[j]
-		return better(loadDetail[id1].LoadPred, loadDetail[id2].LoadPred)
-	})
-	return ids
-}
-
-// Prefer store with larger `count`.
-func selectSrcStoreByCount(loadDetail map[uint64]*storeLoadDetail) uint64 {
-	stores := sortStores(loadDetail, func(lp1, lp2 *storeLoadPred) bool {
-		ld1, ld2 := lp1.min(), lp2.min()
-		if ld1.Count > ld2.Count ||
-			(ld1.Count == ld2.Count && ld1.ByteRate > ld2.ByteRate) {
-			return true
+// smaller is better
+func (bs *balanceSolver) compareSrcStore(st1, st2 uint64) int {
+	if st1 != st2 {
+		// compare source store
+		var lpCmp storeLPCmp
+		if bs.rwTy == write && bs.opTy == transferLeader {
+			lpCmp = sliceLPCmp(
+				minLPCmp(negLoadCmp(sliceLoadCmp(
+					stLdRankCmp(stLdCount, stepRank(bs.maxSrc.Count, bs.rankStep.Count)),
```

Comment:
What the meaning of `rankStep`? it seems always the same in the rank compare.

Reply:
It is used to discretization. Map `float64` to `int64`. For all values in the interval `[rank0 + i * step, rank0 + (i+1) * step)`,  the rank the `i`.

Reply:
Got it. If I understand correctly, the ranking `i` (stepRank return) can only be 0 or 1.  Can this distinguish those stories very well?

Reply:
No, rank  `i = int64((value - rank0) / step)` can be any `int64`.

Reply:
Oh, Sorry. I missed the `step ratio`, can u add some comments about step ratio?

Reply:
Comments added.